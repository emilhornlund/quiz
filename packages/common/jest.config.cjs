// jest.config.cjs
// eslint-disable-next-line no-undef
module.exports = {
  moduleFileExtensions: ['js', 'json', 'ts'],
  rootDir: 'src',
  testRegex: '.*\\.spec\\.ts$',
  transform: {
    '^.+\\.(t|j)s$': [
      'ts-jest',
      { tsconfig: '<rootDir>/../tsconfig.test.json' },
    ],
  },
  collectCoverageFrom: ['**/*.{ts,js}', '!**/*.spec.{ts,js}'],
  coverageDirectory: '../coverage',
  testEnvironment: 'node',
  transformIgnorePatterns: [],
  detectOpenHandles: true,
  forceExit: true,
  maxWorkers: 1,
  setupFilesAfterEnv: ['jest-extended/all'],
  testPathIgnorePatterns: ['/node_modules/', '/dist/'],
}
