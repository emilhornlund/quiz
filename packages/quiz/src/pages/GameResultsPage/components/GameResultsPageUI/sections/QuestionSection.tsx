import {
  GameMode,
  GameResultClassicModeQuestionMetricDto,
  GameResultDto,
  GameResultZeroToOneHundredModeQuestionMetricDto,
} from '@quiz/common'
import React, { FC } from 'react'

import {
  buildQuestionSectionMetricDetails,
  getAveragePrecision,
  getCorrectPercentage,
} from '../utils'

import GameResultTable from './components/GameResultTable'

function getProgress(
  mode: GameMode,
  metric: GameResultDto['questionMetrics'][0],
): number {
  if (mode === GameMode.Classic) {
    return getCorrectPercentage(
      metric as GameResultClassicModeQuestionMetricDto,
    )
  }
  if (mode === GameMode.ZeroToOneHundred) {
    return getAveragePrecision(
      metric as GameResultZeroToOneHundredModeQuestionMetricDto,
    )
  }
  return 0
}

export interface QuestionSectionProps {
  mode: GameMode
  questionMetrics: GameResultDto['questionMetrics']
}

const QuestionSection: FC<QuestionSectionProps> = ({
  mode,
  questionMetrics,
}) => {
  return (
    <section>
      <GameResultTable
        items={questionMetrics.map((metric, index) => ({
          badge: index + 1,
          value: metric.text,
          progress: getProgress(mode, metric),
          details: buildQuestionSectionMetricDetails(mode, metric),
        }))}
      />
    </section>
  )
}

export default QuestionSection
