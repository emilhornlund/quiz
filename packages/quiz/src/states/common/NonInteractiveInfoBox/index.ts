export type { NonInteractiveInfoBoxProps } from './NonInteractiveInfoBox'
export { default } from './NonInteractiveInfoBox'
