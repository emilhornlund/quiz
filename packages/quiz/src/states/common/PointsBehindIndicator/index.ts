export type { PointBehindIndicatorProps } from './PointsBehindIndicator'
export { default } from './PointsBehindIndicator'
