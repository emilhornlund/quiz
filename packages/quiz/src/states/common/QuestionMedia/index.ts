export type { QuestionMediaProps } from './QuestionMedia'
export { default } from './QuestionMedia'
