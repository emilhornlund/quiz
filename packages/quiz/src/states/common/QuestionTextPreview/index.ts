export type { QuestionTextPreviewProps } from './QuestionTextPreview'
export { default } from './QuestionTextPreview'
