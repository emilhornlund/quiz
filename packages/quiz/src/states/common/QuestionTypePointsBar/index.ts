export type { QuestionTypePointsBarProps } from './QuestionTypePointsBar'
export { default } from './QuestionTypePointsBar'
