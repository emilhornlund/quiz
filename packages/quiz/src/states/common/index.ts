export type { NonInteractiveInfoBoxProps } from './NonInteractiveInfoBox'
export type { PointBehindIndicatorProps } from './PointsBehindIndicator'
export type { QuestionAnswerPickerProps } from './QuestionAnswerPicker'
export type { QuestionMediaProps } from './QuestionMedia'
export type { QuestionTextPreviewProps } from './QuestionTextPreview'
export type { QuestionTypePointsBarProps } from './QuestionTypePointsBar'

export { default as GamePage } from './GamePage'
export { default as NonInteractiveInfoBox } from './NonInteractiveInfoBox'
export { default as PointsBehindIndicator } from './PointsBehindIndicator'
export { default as QuestionAnswerPicker } from './QuestionAnswerPicker'
export { default as QuestionMedia } from './QuestionMedia'
export { default as QuestionTextPreview } from './QuestionTextPreview'
export { default as QuestionTypePointsBar } from './QuestionTypePointsBar'
