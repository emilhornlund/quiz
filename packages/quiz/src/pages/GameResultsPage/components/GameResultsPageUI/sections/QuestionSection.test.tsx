import { GameMode, type GameResultDto } from '@quiz/common'
import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import QuestionSection from './QuestionSection'

const h = vi.hoisted(() => {
  return {
    getCorrectPercentage: vi.fn(() => 80),
    getAveragePrecision: vi.fn(() => 47),
    buildQuestionSectionMetricDetails: vi.fn(() => [
      { title: 'Correct', value: '8/10' },
      { title: 'Avg Time', value: '5.1s' },
    ]),
  }
})

vi.mock('../utils', () => ({
  getCorrectPercentage: h.getCorrectPercentage,
  getAveragePrecision: h.getAveragePrecision,
  buildQuestionSectionMetricDetails: h.buildQuestionSectionMetricDetails,
}))

beforeEach(() => {
  h.getCorrectPercentage.mockClear()
  h.getAveragePrecision.mockClear()
  h.buildQuestionSectionMetricDetails.mockClear()
})

describe('QuestionSection', () => {
  it('renders rows with badges and question text for Classic and uses getCorrectPercentage', () => {
    const metrics = [
      { text: 'Q1: Capital of Sweden?' },
      { text: 'Q2: 2 + 2?' },
    ] as unknown as GameResultDto['questionMetrics']

    const { container } = render(
      <QuestionSection mode={GameMode.Classic} questionMetrics={metrics} />,
    )

    const rows = container.querySelectorAll('.tableRow')
    expect(rows.length).toBe(2)
    expect(screen.getByText('1')).toBeInTheDocument()
    expect(screen.getByText('2')).toBeInTheDocument()
    expect(screen.getByText('Q1: Capital of Sweden?')).toBeInTheDocument()
    expect(screen.getByText('Q2: 2 + 2?')).toBeInTheDocument()

    expect(h.getCorrectPercentage).toHaveBeenCalledTimes(2)
    expect(h.getAveragePrecision).not.toHaveBeenCalled()
    expect(h.buildQuestionSectionMetricDetails).toHaveBeenCalledTimes(2)
    expect(h.buildQuestionSectionMetricDetails).toHaveBeenNthCalledWith(
      1,
      GameMode.Classic,
      metrics[0],
    )
    expect(h.buildQuestionSectionMetricDetails).toHaveBeenNthCalledWith(
      2,
      GameMode.Classic,
      metrics[1],
    )

    expect(container).toMatchSnapshot()
  })

  it('toggles details on row click and shows detail items', () => {
    const metrics = [
      { text: 'Q1: Capital of Sweden?' },
    ] as unknown as GameResultDto['questionMetrics']

    const { container } = render(
      <QuestionSection mode={GameMode.Classic} questionMetrics={metrics} />,
    )

    const row = container.querySelector('.tableRow') as HTMLElement
    const details = row.querySelector('.details') as HTMLElement
    expect(details.className).not.toContain('active')

    fireEvent.click(row)
    expect(details.className).toContain('active')
    expect(screen.getByText('Correct')).toBeInTheDocument()
    expect(screen.getByText('8/10')).toBeInTheDocument()
    expect(screen.getByText('Avg Time')).toBeInTheDocument()
    expect(screen.getByText('5.1s')).toBeInTheDocument()

    fireEvent.click(row)
    expect(details.className).not.toContain('active')

    expect(container).toMatchSnapshot()
  })

  it('renders rows for ZeroToOneHundred and uses getAveragePrecision', () => {
    const metrics = [
      { text: 'Q1: Estimate the distance' },
      { text: 'Q2: Estimate the height' },
      { text: 'Q3: Estimate the weight' },
    ] as unknown as GameResultDto['questionMetrics']

    const { container } = render(
      <QuestionSection
        mode={GameMode.ZeroToOneHundred}
        questionMetrics={metrics}
      />,
    )

    const rows = container.querySelectorAll('.tableRow')
    expect(rows.length).toBe(3)
    expect(screen.getByText('Q1: Estimate the distance')).toBeInTheDocument()
    expect(screen.getByText('Q2: Estimate the height')).toBeInTheDocument()
    expect(screen.getByText('Q3: Estimate the weight')).toBeInTheDocument()

    expect(h.getAveragePrecision).toHaveBeenCalledTimes(3)
    expect(h.getCorrectPercentage).not.toHaveBeenCalled()
    expect(h.buildQuestionSectionMetricDetails).toHaveBeenCalledTimes(3)
    expect(h.buildQuestionSectionMetricDetails).toHaveBeenNthCalledWith(
      1,
      GameMode.ZeroToOneHundred,
      metrics[0],
    )

    expect(container).toMatchSnapshot()
  })
})
