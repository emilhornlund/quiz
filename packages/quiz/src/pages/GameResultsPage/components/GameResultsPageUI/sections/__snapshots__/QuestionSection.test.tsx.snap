// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuestionSection > renders rows for ZeroToOneHundred and uses getAveragePrecision 1`] = `
<div>
  <section>
    <div
      class="table"
    >
      <div
        class="tableRow"
      >
        <div
          class="main"
        >
          <div
            class="badge"
          >
            1
          </div>
          <div
            class="value"
            title="Q1: Estimate the distance"
          >
            Q1: Estimate the distance
          </div>
          <div>
            <svg
              class="circularProgressBar correct small"
              height="60"
              width="60"
            >
              <circle
                class="circleBackground"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-width="6"
              />
              <circle
                class="circleProgress"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-linecap="round"
                stroke-width="6"
                style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 89.91238174573988;"
              />
              <text
                class="progressText"
                dominant-baseline="middle"
                text-anchor="middle"
                x="50%"
                y="50%"
              >
                47
                %
              </text>
            </svg>
          </div>
          <div>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down"
              color="black"
              data-icon="chevron-down"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
        <div
          class="details"
        >
          <div
            class="item"
          >
            <div
              class="title"
            >
              Correct
            </div>
            <div
              class="value"
            >
              8/10
            </div>
          </div>
          <div
            class="item"
          >
            <div
              class="title"
            >
              Avg Time
            </div>
            <div
              class="value"
            >
              5.1s
            </div>
          </div>
        </div>
      </div>
      <div
        class="tableRow"
      >
        <div
          class="main"
        >
          <div
            class="badge"
          >
            2
          </div>
          <div
            class="value"
            title="Q2: Estimate the height"
          >
            Q2: Estimate the height
          </div>
          <div>
            <svg
              class="circularProgressBar correct small"
              height="60"
              width="60"
            >
              <circle
                class="circleBackground"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-width="6"
              />
              <circle
                class="circleProgress"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-linecap="round"
                stroke-width="6"
                style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 89.91238174573988;"
              />
              <text
                class="progressText"
                dominant-baseline="middle"
                text-anchor="middle"
                x="50%"
                y="50%"
              >
                47
                %
              </text>
            </svg>
          </div>
          <div>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down"
              color="black"
              data-icon="chevron-down"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
        <div
          class="details"
        >
          <div
            class="item"
          >
            <div
              class="title"
            >
              Correct
            </div>
            <div
              class="value"
            >
              8/10
            </div>
          </div>
          <div
            class="item"
          >
            <div
              class="title"
            >
              Avg Time
            </div>
            <div
              class="value"
            >
              5.1s
            </div>
          </div>
        </div>
      </div>
      <div
        class="tableRow"
      >
        <div
          class="main"
        >
          <div
            class="badge"
          >
            3
          </div>
          <div
            class="value"
            title="Q3: Estimate the weight"
          >
            Q3: Estimate the weight
          </div>
          <div>
            <svg
              class="circularProgressBar correct small"
              height="60"
              width="60"
            >
              <circle
                class="circleBackground"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-width="6"
              />
              <circle
                class="circleProgress"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-linecap="round"
                stroke-width="6"
                style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 89.91238174573988;"
              />
              <text
                class="progressText"
                dominant-baseline="middle"
                text-anchor="middle"
                x="50%"
                y="50%"
              >
                47
                %
              </text>
            </svg>
          </div>
          <div>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down"
              color="black"
              data-icon="chevron-down"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
        <div
          class="details"
        >
          <div
            class="item"
          >
            <div
              class="title"
            >
              Correct
            </div>
            <div
              class="value"
            >
              8/10
            </div>
          </div>
          <div
            class="item"
          >
            <div
              class="title"
            >
              Avg Time
            </div>
            <div
              class="value"
            >
              5.1s
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`QuestionSection > renders rows with badges and question text for Classic and uses getCorrectPercentage 1`] = `
<div>
  <section>
    <div
      class="table"
    >
      <div
        class="tableRow"
      >
        <div
          class="main"
        >
          <div
            class="badge"
          >
            1
          </div>
          <div
            class="value"
            title="Q1: Capital of Sweden?"
          >
            Q1: Capital of Sweden?
          </div>
          <div>
            <svg
              class="circularProgressBar correct small"
              height="60"
              width="60"
            >
              <circle
                class="circleBackground"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-width="6"
              />
              <circle
                class="circleProgress"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-linecap="round"
                stroke-width="6"
                style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 33.92920065876976;"
              />
              <text
                class="progressText"
                dominant-baseline="middle"
                text-anchor="middle"
                x="50%"
                y="50%"
              >
                80
                %
              </text>
            </svg>
          </div>
          <div>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down"
              color="black"
              data-icon="chevron-down"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
        <div
          class="details"
        >
          <div
            class="item"
          >
            <div
              class="title"
            >
              Correct
            </div>
            <div
              class="value"
            >
              8/10
            </div>
          </div>
          <div
            class="item"
          >
            <div
              class="title"
            >
              Avg Time
            </div>
            <div
              class="value"
            >
              5.1s
            </div>
          </div>
        </div>
      </div>
      <div
        class="tableRow"
      >
        <div
          class="main"
        >
          <div
            class="badge"
          >
            2
          </div>
          <div
            class="value"
            title="Q2: 2 + 2?"
          >
            Q2: 2 + 2?
          </div>
          <div>
            <svg
              class="circularProgressBar correct small"
              height="60"
              width="60"
            >
              <circle
                class="circleBackground"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-width="6"
              />
              <circle
                class="circleProgress"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-linecap="round"
                stroke-width="6"
                style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 33.92920065876976;"
              />
              <text
                class="progressText"
                dominant-baseline="middle"
                text-anchor="middle"
                x="50%"
                y="50%"
              >
                80
                %
              </text>
            </svg>
          </div>
          <div>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down"
              color="black"
              data-icon="chevron-down"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
        <div
          class="details"
        >
          <div
            class="item"
          >
            <div
              class="title"
            >
              Correct
            </div>
            <div
              class="value"
            >
              8/10
            </div>
          </div>
          <div
            class="item"
          >
            <div
              class="title"
            >
              Avg Time
            </div>
            <div
              class="value"
            >
              5.1s
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`QuestionSection > toggles details on row click and shows detail items 1`] = `
<div>
  <section>
    <div
      class="table"
    >
      <div
        class="tableRow"
      >
        <div
          class="main"
        >
          <div
            class="badge"
          >
            1
          </div>
          <div
            class="value"
            title="Q1: Capital of Sweden?"
          >
            Q1: Capital of Sweden?
          </div>
          <div>
            <svg
              class="circularProgressBar correct small"
              height="60"
              width="60"
            >
              <circle
                class="circleBackground"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-width="6"
              />
              <circle
                class="circleProgress"
                cx="30"
                cy="30"
                fill="transparent"
                r="27"
                stroke-linecap="round"
                stroke-width="6"
                style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 33.92920065876976;"
              />
              <text
                class="progressText"
                dominant-baseline="middle"
                text-anchor="middle"
                x="50%"
                y="50%"
              >
                80
                %
              </text>
            </svg>
          </div>
          <div>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down"
              color="black"
              data-icon="chevron-down"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
        <div
          class="details"
        >
          <div
            class="item"
          >
            <div
              class="title"
            >
              Correct
            </div>
            <div
              class="value"
            >
              8/10
            </div>
          </div>
          <div
            class="item"
          >
            <div
              class="title"
            >
              Avg Time
            </div>
            <div
              class="value"
            >
              5.1s
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
